src/twisted/internet/unix.py

```diff
@@ -171,10 +171,22 @@ def doRead(self):
             else:
                 return main.CONNECTION_LOST
 
-        if ancillary:
-            ancillaryData = ancillary[0][2]
-            fdCount = len(ancillaryData) // 4
-            fds = struct.unpack('i'*fdCount, ancillaryData)
+        for cmsgLevel, cmsgType, cmsgData in ancillary:
+            if (cmsgLevel != socket.SOL_SOCKET or
+                cmsgType != sendmsg.SCM_RIGHTS):
+                log.msg(
+                    format=(
+                        "%(protocolName)s (on %(hostAddress)r) "
+                        "received unsupported ancillary data "
+                        "(level=%(cmsgLevel)r, type=%(cmsgType)r) "
+                        "from %(peerAddress)r."),
+                    hostAddress=self.getHost(), peerAddress=self.getPeer(),
+                    protocolName=self._getLogPrefix(self.protocol),
+                    cmsgLevel=cmsgLevel, cmsgType=cmsgType,
+                    )
+                continue
+            fdCount = len(cmsgData) // 4
+            fds = struct.unpack('i'*fdCount, cmsgData)
```

Comment:
When working on this my mind also went there: it should be simple to support other types of ancillary data. That's also one of the reasons I initially asked reviewers about factoring out ancillary message decoding.

Having investigated other types of ancillary data I came down to the same conclusion. Other than SCM_RIGHTS and SCM_CREDENTIALS there are a few other obscure ones which tend to be platform-specific.

With that in mind, I decided to:
* Keep the dispatching logic within a single `if .. else` block given that there's no benefit in making it more generic now.
* Factor out the SOL_SOCKET / SCM_RIGHTS decoding and processing to the method you suggested, `_ancillary_SOL_SOCKET_SCM_RIGHTS`. This keeps the code cleaner, points out a way of handling other types of ancillary data and simplifies future dispatching improvements, dict-based or otherwise.